File: day24/src/main.rs
use regex::Regex;
use std::collections::{HashMap, HashSet};

fn main() {
    let input = include_str!("input.txt");
    let mut system = System::new(input);
    assert_eq!(18280, system.fight());
}

#[derive(Debug, Clone, Hash, Eq, PartialEq)]
struct Group {
    id: i32,
    hit_points: i32,
    is_immune: bool,
    weaknesses: Vec<String>,
    immunities: Vec<String>,
    attack_damage: i32,
    attack_type: String,
    initiative: i32,
    units: i32,
}

impl Group {
    fn effective_power(&self) -> i32 {
        self.attack_damage * self.units
    }

    fn effectiveness(&self, other: &Group) -> i32 {
        if other.immunities.contains(&self.attack_type) {
            0
        } else if other.weaknesses.contains(&self.attack_type) {
            2
        } else {
            1
        }
    }

    fn calculate_damage(&self, other: &Group) -> i32 {
        self.effective_power() * self.effectiveness(other)
    }
}

struct System {
    groups: Vec<Group>,
}

impl System {
    fn new(input: &str) -> System {
        let groups = parse(input);
        System { groups }
    }

    fn fight(&mut self) -> i32 {
        let mut unit_by_id = HashMap::new();
        for group in self.groups.clone() {
            unit_by_id.insert(group.id, group);
        }

        loop {
            let (immune_units, infection_units) =
                unit_by_id.iter().fold((0, 0), |groups, (_, group)| {
                    if group.is_immune {
                        (groups.0 + group.units, groups.1)
                    } else {
                        (groups.0, groups.1 + group.units)
                    }
                });
            if immune_units == 0 || infection_units == 0 {
                return immune_units.max(infection_units);
            }
            let mut targets =
                self.target_selection(unit_by_id.values().cloned().collect::<Vec<Group>>());
            targets.sort_by_key(|(id, _)| (!unit_by_id[id].initiative));
            for (attacker, defender) in targets.iter() {
                if !(unit_by_id.contains_key(attacker) && unit_by_id.contains_key(defender)) {
                    continue;
                }
                let attacker = unit_by_id[attacker].clone();
                let mut defender = unit_by_id[defender].clone();
                let damage_dealt = attacker.calculate_damage(&defender);
                let units_killed = damage_dealt / defender.hit_points;
                defender.units = (defender.units - units_killed).max(0);
                if defender.units > 0 {
                    unit_by_id.insert(defender.id, defender);
                } else {
                    unit_by_id.remove(&defender.id);
                }
            }
        }
    }

    fn target_selection(&self, groups: Vec<Group>) -> Vec<(i32, i32)> {
        let mut targets = Vec::new();
        let mut targeted = HashSet::new();

        let mut groups = groups;
        groups.sort_by_key(|u| (!u.effective_power(), !u.initiative));
        for atk in groups.iter() {
            if atk.units <= 0 {
                continue;
            }
            let defender = groups
                .clone()
                .into_iter()
                .filter(|def| {
                    def.is_immune != atk.is_immune
                        && def.units > 0
                        && atk.calculate_damage(def) > 0
                        && !targeted.contains(&def.id)
                })
                .max_by_key(|def| {
                    (
                        atk.calculate_damage(def),
                        def.effective_power(),
                        def.initiative,
                    )
                });
            if let Some(def) = defender {
                targets.push((atk.id, def.id));
                targeted.insert(def.id);
            }
        }
        targets
    }
}

fn parse(input: &str) -> Vec<Group> {
    let units_re = Regex::new(r"^(\d+) units each with (\d+) hit points").unwrap();
    let attack_re =
        Regex::new(r"with an attack that does (\d+) (\w+) damage at initiative (\d+)").unwrap();
    let immune_re = Regex::new(r"\((.+?)\)").unwrap();
    let mut groups = Vec::new();
    let mut id = 0;

    let mut is_immune = true;
    for line in input.trim().lines() {
        if line.starts_with("Immune System:") {
            continue;
        }
        if line.starts_with("Infection:") {
            is_immune = false;
            continue;
        }
        if line.trim().is_empty() {
            continue;
        }

        id += 1;
        let mut group = Group {
            id,
            hit_points: 0,
            is_immune,
            weaknesses: Vec::new(),
            immunities: Vec::new(),
            attack_damage: 0,
            attack_type: String::new(),
            initiative: 0,
            units: 0,
        };

        let units_cap = units_re.captures(line).unwrap();
        let units = units_cap[1].parse::<i32>().unwrap();
        let hit_points = units_cap[2].parse::<i32>().unwrap();
        group.units = units;
        group.hit_points = hit_points;

        if let Some(attack_cap) = attack_re.captures(line) {
            group.attack_damage = attack_cap[1].parse::<i32>().unwrap();
            group.attack_type = attack_cap[2].to_string();
            group.initiative = attack_cap[3].parse::<i32>().unwrap();
        }

        let immune_cap = immune_re.captures(line);
        if immune_cap.is_none() {
            groups.push(group);
            continue;
        }
        let immune_cap = immune_cap.unwrap();
        let mut parts = immune_cap[1].split(';').map(str::trim).collect::<Vec<_>>();
        parts.sort_unstable();
        if parts.len() == 1 {
            if parts[0].starts_with("weak to") {
                let weaknesses = parts[0]
                    .replace("weak to ", "")
                    .trim()
                    .split(',')
                    .map(str::trim)
                    .map(ToString::to_string)
                    .collect();
                group.weaknesses = weaknesses;
            }
            if parts[0].starts_with("immune to") {
                let immunities = parts[0]
                    .replace("immune to ", "")
                    .trim()
                    .split(',')
                    .map(str::trim)
                    .map(ToString::to_string)
                    .collect();
                group.immunities = immunities;
            }
        } else if parts.len() == 2 {
            if parts[0].starts_with("immune to") {
                let immunities = parts[0]
                    .replace("immune to ", "")
                    .trim()
                    .split(',')
                    .map(str::trim)
                    .map(ToString::to_string)
                    .collect();
                group.immunities = immunities;
            }
            if parts[1].starts_with("weak to") {
                let weaknesses = parts[1]
                    .replace("weak to ", "")
                    .trim()
                    .split(',')
                    .map(str::trim)
                    .map(ToString::to_string)
                    .collect();
                group.weaknesses = weaknesses;
            }
        }
        groups.push(group);
    }
    groups
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let input = "Immune System:
17 units each with 5390 hit points (weak to radiation, bludgeoning) with an attack that does 4507 fire damage at initiative 2
989 units each with 1274 hit points (immune to fire; weak to bludgeoning, slashing) with an attack that does 25 slashing damage at initiative 3

Infection:
801 units each with 4706 hit points (weak to radiation) with an attack that does 116 bludgeoning damage at initiative 1
4485 units each with 2961 hit points (immune to radiation; weak to fire, cold) with an attack that does 12 slashing damage at initiative 4";

        let mut system = System::new(input);
        assert_eq!(5216, system.fight());
    }
}

File: day24/main.js
const example = `Immune System:
17 units each with 5390 hit points (weak to radiation, bludgeoning) with an attack that does 4507 fire damage at initiative 2
989 units each with 1274 hit points (immune to fire; weak to bludgeoning, slashing) with an attack that does 25 slashing damage at initiative 3

Infection:
801 units each with 4706 hit points (weak to radiation) with an attack that does 116 bludgeoning damage at initiative 1
4485 units each with 2961 hit points (immune to radiation; weak to fire, cold) with an attack that does 12 slashing damage at initiative 4`;

const input = `Immune System:
554 units each with 8034 hit points (weak to cold; immune to slashing) with an attack that does 124 bludgeoning damage at initiative 2
285 units each with 3942 hit points (weak to cold) with an attack that does 107 bludgeoning damage at initiative 6
4470 units each with 7895 hit points (immune to bludgeoning; weak to radiation) with an attack that does 17 bludgeoning damage at initiative 1
4705 units each with 8128 hit points (weak to slashing) with an attack that does 14 bludgeoning damage at initiative 8
3788 units each with 7504 hit points (weak to cold, slashing) with an attack that does 17 cold damage at initiative 3
7087 units each with 2733 hit points (weak to bludgeoning) with an attack that does 3 slashing damage at initiative 14
23 units each with 7234 hit points with an attack that does 3132 fire damage at initiative 15
818 units each with 7188 hit points (weak to fire; immune to radiation, slashing) with an attack that does 80 fire damage at initiative 19
3233 units each with 3713 hit points (immune to cold; weak to radiation) with an attack that does 10 radiation damage at initiative 20
1011 units each with 8135 hit points (immune to slashing, cold, bludgeoning; weak to fire) with an attack that does 75 radiation damage at initiative 12

Infection:
136 units each with 37513 hit points (weak to radiation) with an attack that does 492 cold damage at initiative 18
4811 units each with 5863 hit points (weak to radiation, cold; immune to slashing) with an attack that does 2 radiation damage at initiative 17
4057 units each with 9812 hit points (weak to slashing) with an attack that does 4 bludgeoning damage at initiative 11
2828 units each with 30926 hit points (immune to bludgeoning; weak to cold) with an attack that does 19 cold damage at initiative 7
2311 units each with 20627 hit points (immune to slashing) with an attack that does 17 slashing damage at initiative 5
1622 units each with 30824 hit points (weak to slashing, bludgeoning) with an attack that does 34 bludgeoning damage at initiative 4
108 units each with 8628 hit points with an attack that does 139 slashing damage at initiative 13
1256 units each with 51819 hit points (immune to cold, slashing) with an attack that does 63 radiation damage at initiative 16
3681 units each with 21469 hit points (weak to slashing; immune to cold, bludgeoning) with an attack that does 11 cold damage at initiative 9
7289 units each with 6935 hit points (weak to slashing, bludgeoning) with an attack that does 1 fire damage at initiative 10`;

function parseRow(line) {
  const matches = [
    ...line.matchAll(
      /(?<units>\d+) units each with (?<hitPoints>\d+) hit points/gi
    ),
  ];
  const units = Number(matches?.[0]?.groups?.units);
  const hitPoints = Number(matches?.[0]?.groups?.hitPoints);

  const strengthWeaknessesMatches = [
    ...line.matchAll(/\((?<strengthWeaknesses>.+)\)/gi),
  ];
  const strengthWeaknesses =
    strengthWeaknessesMatches?.[0]?.groups?.strengthWeaknesses ?? "";
  const parts = strengthWeaknesses
    .split(";")
    .map((s) => s.trim())
    .sort();
  const immuneTo = [];
  const weakAgainst = [];
  for (let part of parts) {
    if (part.startsWith("weak to")) {
      weakAgainst.push(
        ...part
          .replaceAll("weak to", "")
          .split(",")
          .map((s) => s.trim())
      );
    }
    if (part.startsWith("immune to")) {
      immuneTo.push(
        ...part
          .replaceAll("immune to", "")
          .split(",")
          .map((s) => s.trim())
      );
    }
  }
  const attackMatches = [
    ...line.matchAll(
      /with an attack that does (?<damage>\d+) (?<attackType>\w+) damage at initiative (?<initiative>\d+)/gi
    ),
  ];
  const damage = Number(attackMatches?.[0]?.groups?.damage);
  const attackType = attackMatches?.[0]?.groups?.attackType;
  const initiative = Number(attackMatches?.[0]?.groups?.initiative);

  return {
    units,
    hitPoints,
    immuneTo,
    weakAgainst,
    damage,
    attackType,
    initiative,
  };
}

function parse(input) {
  const lines = input.trim().split("\n");
  let group = "IS";
  const groups = [];
  let id = 0;
  for (const line of lines) {
    if (line.startsWith("Immune System:")) {
      group = "IS";
      continue;
    }

    if (line.startsWith("Infection:")) {
      group = "I";
      continue;
    }

    if (line.trim() === "") {
      continue;
    }
    id++;
    const units = parseRow(line);
    units.group = group;
    units.id = id;
    groups.push(units);
  }
  return groups;
}

function effectivePower({ units, damage }) {
  return units * damage;
}

function damage(attacker, defender) {
  if (defender.weakAgainst.includes(attacker.attackType)) {
    return effectivePower(attacker) * 2;
  }
  if (defender.immuneTo.includes(attacker.attackType)) {
    return 0;
  }
  return effectivePower(attacker);
}

function targetSelection(groups = []) {
  // Target selection.
  groups.sort((a, b) => {
    const aep = effectivePower(a);
    const bep = effectivePower(b);
    const sameEffectivePower = bep === aep;
    const initiativeDesc = b.initiative - a.initiative;
    const effectivePowerDesc = bep - aep;
    return sameEffectivePower ? initiativeDesc : effectivePowerDesc;
  });

  const pairs = {};
  const selected = new Set();
  for (const group of groups) {
    if (group.units <= 0) {
      continue;
    }
    const others = groups.filter((other) => {
      return (
        other.group != group.group &&
        other.units > 0 &&
        damage(group, other) > 0 &&
        !selected.has(other.id)
      );
    });
    others.sort((a, b) => {
      const admg = damage(group, a);
      const bdmg = damage(group, b);

      const aep = effectivePower(a);
      const bep = effectivePower(b);

      return bdmg === admg
        ? bep === aep
          ? b.initiative - a.initiative
          : bep - aep
        : bdmg - admg;
    });
    if (others.length) {
      pairs[group.id] = others[0].id;
      selected.add(others[0].id);
    }
  }

  return Object.entries(pairs).map((kv) => kv.map(Number));
}

function fight(groupById = {}, pairs = []) {
  pairs.sort(([a], [b]) => groupById[b].initiative - groupById[a].initiative);

  for (const [attackerId, defenderId] of pairs) {
    const attacker = groupById[attackerId];
    const defender = groupById[defenderId];
    if (!(attacker && defender)) {
      continue;
    }
    const damageDealt = damage(attacker, defender);
    const unitsDie = Math.floor(damageDealt / defender.hitPoints);
    defender.units -= unitsDie;
    defender.units = Math.max(0, defender.units);
    if (!defender.units) {
      delete groupById[defender.id];
    } else {
      groupById[defender.id] = defender;
    }
  }

  return Object.values(groupById);
}

function hasGroup(groups = []) {
  const unitsByGroup = groups.reduce((acc, group) => {
    if (!acc[group.group]) {
      acc[group.group] = 0;
    }
    acc[group.group] += group.units;
    return acc;
  }, {});
  return Object.keys(unitsByGroup).length > 1;
}

function boostDamage(groups = [], boost = 0) {
  return groups.map((group) => {
    if (group.group === "IS") {
      group.damage += boost;
    }
    return group;
  });
}

function execute(input, boost = 0) {
  let groups = boostDamage(parse(input), boost);
  while (hasGroup(groups)) {
    const unitsBefore = groups.reduce((acc, group) => acc + group.units, 0);
    const pairs = targetSelection(groups);
    const groupById = groups.reduce((acc, group) => {
      acc[group.id] = group;
      return acc;
    }, {});
    groups = fight(groupById, pairs);
    const unitsAfter = groups.reduce((acc, group) => acc + group.units, 0);
    if (unitsBefore === unitsAfter) {
      return groups;
    }
  }
  return groups;
}

function part1(input) {
  const groups = execute(input);
  const unitsRemaining = groups.reduce((acc, group) => acc + group.units, 0);
  return unitsRemaining;
}

function immuneSystemWin(groups = []) {
  return groups.every((group) => group.group === "IS");
}

function part2(input) {
  let boost = 0;
  groups = execute(input, boost);
  while (!immuneSystemWin(groups)) {
    boost += 1;
    groups = execute(input, boost);
  }
  const unitsRemaining = groups.reduce((acc, group) => acc + group.units, 0);
  console.log("boost: " + boost, "units", unitsRemaining);
  return unitsRemaining;
}

(function main() {
  if (part1(example) !== 5216) {
    throw new Error("part1Error: got " + part1(example));
  }
  if (part1(input) !== 18280) {
    throw new Error("part1Error: got " + part1(input));
  }

  if (part2(example) !== 51) {
    throw new Error("part2Error: got " + part2(example));
  }

  if (part2(input) !== 4573) {
    throw new Error("part2Error: got " + part2(input));
  }
})();
